src/QirRuntime/lib/Tracer/tracer.hpp

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <limits>
+#include <memory>
+#include <ostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "CoreTypes.hpp"
+#include "QuantumApi_I.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    using OpId = int;
+    using Time = int;
+    using Duration = int;
+    using LayerId = size_t;
+
+    constexpr LayerId INVALID = std::numeric_limits<size_t>::max();
+
+    /*==================================================================================================================
+        Layer
+    ==================================================================================================================*/
+    struct Layer
+    {
+        // Width of the layer on the time axis.
+        const Duration duration;
+
+        // Start time of the layer.
+        const Time startTime;
+
+        // Quantum operations, assigned to this layer.
+        std::unordered_map<OpId, int /*count of the op with this id*/> operations;
+
+        // Optional id, if the layer represents a global barrier.
+        OpId barrierId = -1;
+
+        Layer(Duration duration, Time startTime)
+            : duration(duration)
+            , startTime(startTime)
+        {
+        }
+    };
+
+    /*==================================================================================================================
+        QubitState
+    ==================================================================================================================*/
+    struct QubitState
+    {
+        // The last layer this qubit was used in, `INVALID` means the qubit haven't been used yet in any
+        // operations of non-zero duration.
+        LayerId layer = INVALID;
+
+        // For layers with duration greater than one, multiple operations might fit on the same qubit, if the operations
+        // are short. `lastUsedTime` is the end time of the last operation, the qubit participated it.

```

Comment:
multiple operations **on the same qubit** might fit **in the layer**?
the qubit participated ~~it~~ **in** ?

Reply:
Reworded.